internal/core/software_renderer.rs

```diff
@@ -1667,6 +1731,18 @@ impl<'a, T: ProcessScene> crate::item_rendering::ItemRenderer for SceneBuilder<'
                 self.draw_text_paragraph(paragraph, physical_clip, offset, color);
             }
         }
+
+        let cursor_rect = self
+            .window
+            .window_adapter()
+            .renderer()
+            .text_input_cursor_rect_for_byte_offset(
```

Comment:
This shapes the text a second time, doing the same work twice. This is sub-optimal.

`TextParagraphLayout` could have an `lines: Option<Vec<TextLine>>` and have a `cache_layout()` function to populate it. If available, then `layout_lines()` could use that instead of computing the lines on the fly.

This could be done in a follow-up change though.